### PR TITLE
Added deprecation to cwd not existing Fixes #18249

### LINKED ISSUE #18249: [Process] Silent behaviour change if folder doesn't exist
In Process, the second parameter is the `cwd`. If this folder doesn't exist, it seems to run a getcwd.

This should initiate / throw a warning or exception, since the behaviour change is / can be critical.

For us, we were performing git clone into a folder, but the folder doesn't exist, so it clones into the folder where console is ran from. This is on ~3.0

### PR DESCRIPTION
**Pull request by alexbowers**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  |no 
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #18249
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 1]
```diff
@@ -47,6 +47,27 @@ protected function tearDown()
         }
     }
 
+    /**
+     * @see https://github.com/symfony/symfony/issues/18249
+     */
+    public function testCwdParameterBugDeprecationIssue18249()
```
Comment (chalasr) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: just `testInvalidCwd`

**Comment by alexbowers** [comment]
@chalasr I have renamed the test

File: UPGRADE-3.4.md [hunk 2]
```diff
@@ -68,6 +68,9 @@ Process
 
  * The `Symfony\Component\Process\ProcessBuilder` class has been deprecated,
    use the `Symfony\Component\Process\Process` class directly instead.
+   
+ * The `Symfony\Component\Process\Process` __construct `cwd` argument will throw 
+   a deprecation warning on `run` if the folder does not exist.
```
Comment (chalasr) [inline UPGRADE-3.4.md]: we should replace the (silent) deprecation log by a real exception in 4.0, `UPGRADE-4.0.md` should mention that

**Review by chalasr — CHANGES_REQUESTED** [review]

File: UPGRADE-3.4.md (hunk 2 above)
Comment (alexbowers) [inline UPGRADE-3.4.md]: Should I put that in the upgrade log before the exception is even implemented? Or should there be a separate PR for the exception?

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 3]
```diff
@@ -47,6 +47,27 @@ protected function tearDown()
...
+    {
+        // Check that it works fine if the CWD exists
+        $cmd = new Process('touch testing1.txt', __DIR__);
+        $this->assertEquals(0, $cmd->run());
+
+        try {
+            $cmd = new Process('touch testing2.txt', __DIR__.'/notfound/');
+            $this->assertEquals(0, $cmd->run());
+        } catch (\Throwable $e) {
+            $this->assertTrue(true); // A deprecated error was thrown
```
Comment (alexbowers) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: This has been done

File: UPGRADE-3.4.md (hunk 2 above)
Comment (chalasr) [inline UPGRADE-3.4.md]: this should target `3.4`. The upgrade note should be added when deprecating the behavior, we take care of doing the change when merging it up to master (because master is 4.0, where we remove BC layers).

Comment (alexbowers) [inline UPGRADE-3.4.md]: How do i change the target?

If i change it now, surely doesn't it put all of `master` onto ` 3.4` which isn't what we want.

Comment (robfrawley) [inline UPGRADE-3.4.md]: You need to rebase your chases on top of `3.4` and then you can change the base branch by clicking the "Edit" button on the top right of your PR.

Comment (alexbowers) [inline UPGRADE-3.4.md]: Any chance you can give me the details on how to rebase?

Comment (alexbowers) [inline UPGRADE-3.4.md]: Alternatively, I'm going away for the weekend. If somebody else wants to apply these as a patch to 3.4 instead of master, i'm more than happy for you to go ahead and do it.

Comment (robfrawley) [inline UPGRADE-3.4.md]: https://git-scm.com/docs/git-rebase

File: src/Symfony/Component/Process/Process.php [hunk 4]
```diff
@@ -303,6 +303,10 @@ public function start(callable $callback = null, array $env = array())
             $ptsWorkaround = fopen(__FILE__, 'r');
         }
 
+        if (!file_exists($this->cwd)) {
```
Comment (fabpot) [inline src/Symfony/Component/Process/Process.php]: Shoudn't it be `is_dir()`?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 5]
```diff
@@ -47,6 +47,25 @@ protected function tearDown()
         }
     }
 
+    /**
+     * @see https://github.com/symfony/symfony/issues/18249
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: Usually we don't add such notes, we rely on the git history instead, should be removed.

File: src/Symfony/Component/Process/Process.php [hunk 6]
```diff
@@ -303,6 +303,10 @@ public function start(callable $callback = null, array $env = array())
...
+            @trigger_error('The provided cwd does not exist. Command is currently ran against getcwd()', E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: the message should mention that an exception will be thrown in 4.0

Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: oh, and the message should tell since which version this deprecation exist (see other deprecation messages in the code)

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 7]
```diff
@@ -47,6 +47,25 @@ protected function tearDown()
...
+     *
+     * @group legacy
+     * @expectedDeprecation The provided cwd does not exist. Command is currently ran against getcwd()
+     */
+    public function testInvalidCwd()
+    {
+        // Check that it works fine if the CWD exists
+        $cmd = new Process('touch testing1.txt', __DIR__);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: fails on Windows

**Comment by alexbowers** [comment]
I'll fix the other issues either his evening or tomorrow, but I don't have access to a windows machine to test or find a working test. Thanks

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by alexbowers** [comment]
@nicolas-grekas Any chance you can provide a fix for windows? I have no machine to test it on.

I will fix the other issues.

**Comment by nicolas-grekas** [comment]
It looks like Windows already fails when the cwd doesn't exist:
https://msdn.microsoft.com/en-us/library/ms681382%28v=vs.85%29.aspx#ERROR_DIRECTORY

So you can just throw an exception on Windows - and adjust the test accordingly.

**Comment by nicolas-grekas** [comment]
ping @alexbowers

**Comment by alexbowers** [comment]
Thanks, I'll try to get around to it this evening. Completely forgot about this.

File: UPGRADE-3.4.md (hunk 2 above)
Comment (alexbowers) [reply UPGRADE-3.4.md]: I believe this has been done properly now.

File: src/Symfony/Component/Process/Process.php (hunk 4 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Done.

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 5 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Removed.

File: src/Symfony/Component/Process/Process.php (hunk 6 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Done.

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 7 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: This should hopefully be fixed now.

**Comment by alexbowers** [comment]
I think all outstanding issues were resolved, can this be confirmed?

File: src/Symfony/Component/Process/Process.php [hunk 8]
```diff
@@ -334,7 +337,11 @@ public function start(callable $callback = null/*, array $env = array()*/)
             $ptsWorkaround = fopen(__FILE__, 'r');
         }
 
-        $this->process = proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $env, $this->options);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: the $this->options property must be used, not doing so breaks windows badly

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 9]
```diff
@@ -48,6 +48,23 @@ protected function tearDown()
...
     }
 
+    /**
+     * @group legacy
+     * @expectedDeprecation The provided cwd does not exist. Command is currently ran against getcwd(). This behaviour is deprecated since version 3.4 and will be removed in 4.0.
+     */
+    public function testInvalidCwd()
+    {
+        // Check that it works fine if the CWD exists
+        $cmd = new Process('echo test > testing1.txt', __DIR__);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: there is no need to pipe the output to a file

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 10]
```diff
@@ -48,6 +48,23 @@ protected function tearDown()
...
+        $this->assertEquals(0, $cmd->run());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: should be `assertSame` - but to me this assertion is not needed: the fact that a process runs correctly when all is fine is already tested in other cases

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 11]
```diff
@@ -48,6 +48,23 @@ protected function tearDown()
...
+
+        $cmd = new Process('echo test > testing2.txt', __DIR__.'/notfound/');
+        $this->assertEquals(0, $cmd->run());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: this fails on windows, where suppress_error does make it fallback to getcwd (but fail instead)

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 9 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Done

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 10 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Done

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 11 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: How should I handle this?

I cannot test on windows, because I do not have a device to test it with.

**Comment by alexbowers** [comment]
@chalasr @fabpot @nicolas-grekas 

I believe all changes have been implemented.

File: UPGRADE-3.4.md [hunk 12]
```diff
@@ -68,6 +68,9 @@ Process
 
  * The `Symfony\Component\Process\ProcessBuilder` class has been deprecated,
    use the `Symfony\Component\Process\Process` class directly instead.
+   
+ * The `Symfony\Component\Process\Process` __construct `cwd` argument will throw
```
Comment (nicolas-grekas) [inline UPGRADE-3.4.md]: [...] constructor's `$cwd` argument [...]

File: src/Symfony/Component/Process/Process.php [hunk 13]
```diff
@@ -318,6 +318,7 @@ public function start(callable $callback = null/*, array $env = array()*/)
         } elseif (null !== $env) {
             @trigger_error('Not inheriting environment variables is deprecated since Symfony 3.3 and will always happen in 4.0. Set "Process::inheritEnvironmentVariables()" to true instead.', E_USER_DEPRECATED);
         }
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: should be reverted

File: src/Symfony/Component/Process/Process.php [hunk 14]
```diff
@@ -334,6 +335,10 @@ public function start(callable $callback = null/*, array $env = array()*/)
             $ptsWorkaround = fopen(__FILE__, 'r');
         }
 
+        if (!is_dir($this->cwd)) {
+            @trigger_error('The provided cwd does not exist. Command is currently ran against getcwd(). This behaviour is deprecated since version 3.4 and will be removed in 4.0.', E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php]: I think we should stick to my previous suggestion: on Windows, we should throw an exception instead of a deprecation, because on Windows, it's already broken, the getcwd fallback does not apply.

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 15]
```diff
@@ -48,6 +48,43 @@ protected function tearDown()
         }
     }
 
+    /**
+     * @group legacy
+     * @expectedDeprecation The provided cwd does not exist. Command is currently ran against getcwd(). This behaviour is deprecated since version 3.4 and will be removed in 4.0.
+     */
+    public function testInvalidCwd()
+    {
+        if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: should be `if ('\\' === DIRECTORY_SEPARATOR) {` to test for Windows, as everywhere else in the code base.

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 16]
```diff
@@ -48,6 +48,43 @@ protected function tearDown()
...
+        $cmd->run();
+
+        $cmd = new Process('echo test', __DIR__.'/notfound/');
+        $cmd->run();
+    }
+
+    /**
+     * @group legacy
+     * @expectedDeprecation The provided cwd does not exist. Command is currently ran against getcwd(). This behaviour is deprecated since version 3.4 and will be removed in 4.0.
+     * @expectedException \PHPUnit_Framework_Error_Warning
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: should be replaced by an `@expectedException`+`@expectedExceptionMessage` as suggested above

**Review by nicolas-grekas — COMMENTED** [review]
almost good :)

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 16 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Done

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 15 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Done

File: src/Symfony/Component/Process/Process.php (hunk 14 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Done

File: UPGRADE-3.4.md (hunk 12 above)
Comment (alexbowers) [reply UPGRADE-3.4.md]: Done

**Review by nicolas-grekas — APPROVED** [review]

File: UPGRADE-3.4.md [hunk 17]
```diff
@@ -69,6 +69,9 @@ Process
  * The `Symfony\Component\Process\ProcessBuilder` class has been deprecated,
    use the `Symfony\Component\Process\Process` class directly instead.
 
+ * The `Symfony\Component\Process\Process` constructor's $cwd argument will throw
```
Comment (xabbuh) [inline UPGRADE-3.4.md]: `$cwd` should also be enclosed with backticks

File: src/Symfony/Component/Process/Process.php [hunk 18]
```diff
@@ -334,6 +334,14 @@ public function start(callable $callback = null/*, array $env = array()*/)
             $ptsWorkaround = fopen(__FILE__, 'r');
         }
 
+        if (!is_dir($this->cwd)) {
+            if ('\\' === DIRECTORY_SEPARATOR) {
+                throw new RuntimeException('The provided cwd does not exist.');
+            }
+
+            @trigger_error('The provided cwd does not exist. Command is currently ran against getcwd(). This behaviour is deprecated since version 3.4 and will be removed in 4.0.', E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/Process/Process.php]: behavior

File: UPGRADE-3.4.md (hunk 17 above)
Comment (xabbuh) [reply UPGRADE-3.4.md]: I think we should actually reword this (please also add the same entry to the `CHANGELOG.md` file of the `Process` component):

> Passing a not existing working directory to the constructor of the `Symfony\Component\Process\Process` class is deprecated and will not be supported anymore in Symfony 4.0.

And please also add a similar entry to the `UPGRADE-4.0.md` file:

> Passing a not existing working directory to the constructor of the `Symfony\Component\Process\Process` class is not supported anymore.

File: src/Symfony/Component/Process/Process.php (hunk 18 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Does Symfony follow british or american spelling across the board?

Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Process.php]: US

Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Ok, ill make that change this evening

File: UPGRADE-3.4.md (hunk 17 above)
Comment (alexbowers) [reply UPGRADE-3.4.md]: Are these messages verbatim?

File: src/Symfony/Component/Process/Process.php (hunk 18 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Process.php]: Done

File: UPGRADE-3.4.md (hunk 17 above)
Comment (alexbowers) [reply UPGRADE-3.4.md]: Done

**Comment by alexbowers** [comment]
@fabpot @nicolas-grekas @chalasr 

Ready for review.

**Comment by alexbowers** [comment]
@fabpot do I resolve these conflicts, or do they get resolved upstream?

**Comment by nicolas-grekas** [comment]
Can you rebase to resolve the conflicts please? (Don't use the GitHub UI, there should be no merge commit.)

**Comment by alexbowers** [comment]
Whenever I try to rebase, I have no conflicts.

This is saying just one conflict. What am I doing wrong?
```
git rebase origin/3.4 -i

On branch issue/process-cwd-bug
Your branch is up-to-date with 'github/issue/process-cwd-bug'.
nothing to commit, working tree clean
```

**Comment by chalasr** [comment]
You should have an upstream remote to rebase against. Assuming `origin` is your fork:

```
git remote add upstream git@github.com:symfony/symfony.git
git fetch upstream
git rebase upstream/3.4
```

**Comment by alexbowers** [comment]
`origin` is my upstream. I have `origin` and `github` as remotes.

`origin` is this repo, `github` is my fork.

**Comment by alexbowers** [comment]
Nevermind. It appears my GUI lies to me.

`git remote -v` shows they both point to my fork.

My gui shows differently. Strange.

**Comment by alexbowers** [comment]
So, uh. I think i fucked up.

**Comment by alexbowers** [comment]
Why would this have caused these issues? 

image

**Comment by alexbowers** [comment]
Should this just be a new PR with the commits cherrypicked onto the right base?

**Comment by nicolas-grekas** [comment]
You should rebase **-i**, then keep only your commit if there are more

**Comment by alexbowers** [comment]
Oh. How do i fix this?

**Comment by nicolas-grekas** [comment]
Oh, and you should not "commit" when resolving a conflict, but "rebase --continue" instead.

**Comment by alexbowers** [comment]
Is it possible to fix what i've done?

**Comment by chalasr** [comment]
@alexbowers if you use `git reflog`, you'll see your local history e.g:
```
36139ca HEAD@{23}: rebase finished: returning to refs/heads/master
36139ca HEAD@{24}: rebase: checkout refs/remotes/origin/masterefs/remotes/origin/master
684c07c HEAD@{25}: checkout: moving from 2.x to master
```
Find the best entry to return just at the state just before the mistake happened, then hard reset on it `git reset --hard HEAD@{25}` (assuming `HEAD@{25}` is the best fit).
Once you did this you can start rebasing.

**Comment by alexbowers** [comment]
```
➜  symfony git:(issue/process-cwd-bug) git rebase -i upstream/3.4
Successfully rebased and updated refs/heads/issue/process-cwd-bug.
➜  symfony git:(issue/process-cwd-bug) git status
On branch issue/process-cwd-bug
Your branch and 'github/issue/process-cwd-bug' have diverged,
and have 20 and 41 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working tree clean
```

That seems to have worked. But it says I would have to pull to get the changes.

Am i still doing something wrong here?

**Comment by alexbowers** [comment]
I also didn't get a conflict that time for some reason. My GUI is also saying i'm 541 commits behind.

**Comment by robfrawley** [comment]
> Should this just be a new PR with the commits cherry-picked onto the right base?

FYI: You don't need to create a new PR to cherry pick your commit(s) onto a fresh branch if that's what you're comfortable with. Just create a new branch based on `3.4` and cherry pick the commits you want for this PR, then delete this branch and make a new branch with the same name. Then just force push (`git push --force`). Not that this is the best approach, but if rebasing is something you aren't comfortable with and cherry picking is, the option exists.

**Comment by alexbowers** [comment]
I think i've found my way back to sanity.

I _think_ its okay now.

**Comment by alexbowers** [comment]
The codestyle patch https://fabbot.io/patch/symfony/symfony/23708/b93e7ad5cb0dc7ae9244bc1674044e4bbd379d56/cs.diff is not what i've modified. Is this from other PRs?

**Review by chalasr — APPROVED** [review]

File: UPGRADE-4.0.md [hunk 19]
```diff
@@ -566,6 +566,7 @@ Ldap
 
 Process
 -------
+ * Passing a not existing working directory to the constructor of the `Symfony\Component\Process\Process` class is not supported anymore.
```
Comment (nicolas-grekas) [inline UPGRADE-4.0.md]: Missing blank line

**Review by nicolas-grekas — APPROVED** [review]

File: UPGRADE-4.0.md (hunk 19 above)
Comment (alexbowers) [reply UPGRADE-4.0.md]: Fixed.

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 20]
```diff
@@ -48,6 +48,42 @@ protected function tearDown()
...
+     * @expectedException \Symfony\Component\Process\Exception\RuntimeException
+     * @expectedExceptionMessage The provided cwd does not exist.
+     */
+    public function testInvalidCwdOnWindows()
+    {
+        if ('\\' !== DIRECTORY_SEPARATOR) {
+            $this->markTestSkipped('Unix handles this automatically.');
+        }
+
+        // Check that it works fine if the CWD exists
```
Comment (stof) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: This does not test anything. To test that it works, it must be a separate test.

If the exception happens in the first process rather than the second one, the `@expectedException` assertion will be fine with it.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Process/Tests/ProcessTest.php (hunk 20 above)
Comment (alexbowers) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Handled this by forcing a fail if any exception occurs in that first section.

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 21]
```diff
@@ -1494,7 +1534,7 @@ public function testRawCommandLine()
 (
     [0] => -
     [1] => a
-    [2] => 
+    [2] =>
```
Comment (chalasr) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: should be reverted (breaks tests)

File: src/Symfony/Component/Process/CHANGELOG.md [hunk 22]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 -----
 
  * deprecated the ProcessBuilder class
+ * Passing a not existing working directory to the constructor of the `Symfony\Component\Process\Process` class is deprecated and will not be supported anymore in Symfony 4.0.
```
Comment (chalasr) [inline src/Symfony/Component/Process/CHANGELOG.md]:
```
 * deprecated calling `Process::start()` without setting a valid working directory beforehand (via `setWorkingDirectory()` or constructor)
```

File: UPGRADE-3.4.md [hunk 23]
```diff
@@ -210,6 +210,8 @@ Process
  * The `Symfony\Component\Process\ProcessBuilder` class has been deprecated,
    use the `Symfony\Component\Process\Process` class directly instead.
 
+ * Passing a not existing working directory to the constructor of the `Symfony\Component\Process\Process` class is deprecated and will not be supported anymore in Symfony 4.0.
```
Comment (chalasr) [inline UPGRADE-3.4.md]: Actually `Process::__construct()` allows `null` for `$cwd` (and fills the property using `getcwd()`), allowing to call `setWorkingDirectory()` later. Unless we want to make it non nullable, this should talk about `start()`, not the constructor. 
I propose:

> Calling `Process::start()` without setting a valid working directory (via `setWorkingDirectory()` or constructor) beforehand is deprecated and will throw an exception in 4.0.

**Review by chalasr — CHANGES_REQUESTED** [review]

**Comment by nicolas-grekas** [comment]
@alexbowers anytime this week to fix the remaining comments?

**Comment by alexbowers** [comment]
Yes, I'll put it in my todo list.

**Comment by nicolas-grekas** [comment]
@alexbowers don't forget this PR, we're almost at it and it has to be in 3.4 :)

**Comment by alexbowers** [comment]
@nicolas-grekas Sent the patch again.

**Review by chalasr — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Can you just apply fabbot's patch please?

**Comment by alexbowers** [comment]
Done, though that is for changes not related to anything i've done as far as I can tell.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexbowers.

**Comment by alexbowers** [comment]
🍾 Only took me 69 days to actually get around to doing this.

**Comment by fabpot** [comment]
Congrats! And thank you for your work and persistence @alexbowers, this is much appreciated.
